### PR TITLE
XDP UT framework

### PR DESCRIPTION
**Pull request by mazdakn**
## Description

This PR allows to run unit tests in XDP mode. List of changes are:
- In bpf/ut directory a flag called "forXDP" is added to differentiate tc and xdp test setup. For example, to use correct arguments in bpftool.

## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests (delete as appropriate) In plan/Not needed/Done

## Release Note

```release-note
None required
```

### CODE REVIEW

File: bpf/ut/ip_dec_ttl_test.go [hunk 1]
```diff
@@ -23,10 +23,38 @@ import (
...
+
+		ip35 := *ipv4Default
+		ip35.TTL = 127
+		_, _, _, _, pktBytes, err = testPacket(nil, &ip35, nil, nil)
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(res.dataOut).To(Equal(pktBytes))
+	})
+}
+
```
Comment (mazdakn) [inline bpf/ut/ip_dec_ttl_test.go:53]: This test does not make sense for XDP. It is just here to test the UT framework. Will be removed in the end.

File: bpf-gpl/ut/ip_dec_ttl.xdp.c [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	if (skb_refresh_validate_ptrs(ctx, UDP_SIZE)) {
+		ctx->fwd.reason = CALI_REASON_SHORT;
+		CALI_DEBUG("Too short\n");
+		return -1;
+	}
+
+	ip_dec_ttl(ctx->ip_header);
+
+	return 0;
+}
```
Comment (mazdakn) [inline bpf-gpl/ut/ip_dec_ttl.xdp.c:31]: This file is only to test ip_dec_ttl in xdp mode, just to test the xdp ut framework. Will be removed in the end.

File: bpf/ut/bpf_prog_test.go [hunk 3]
```diff
@@ -309,12 +311,13 @@ func initMapsOnce() {
 		ipsMap = ipsets.Map(mc)
 		stateMap = state.Map(mc)
 		testStateMap = state.MapForTest(mc)
-		jumpMap = jump.MapForTest(mc)
+		xdpJumpMap = MapForTest(mc)
```
Comment (neiljerram) [inline bpf/ut/bpf_prog_test.go:314]: nit: maybe move into its own package, e.g. "xdpjump" ?

File: bpf/ut/icmp_port_unreachable_test.go [hunk 4]
```diff
@@ -31,7 +31,7 @@ func TestICMPPortUnreachable(t *testing.T) {
 	_, ipv4, _, _, pktBytes, err := testPacketUDPDefault()
 	Expect(err).NotTo(HaveOccurred())
 
-	runBpfUnitTest(t, "icmp_port_unreachable.c", func(bpfrun bpfProgRunFn) {
+	runBpfUnitTest(t, "icmp_port_unreachable.c", false, func(bpfrun bpfProgRunFn) {
```
Comment (neiljerram) [inline bpf/ut/icmp_port_unreachable_test.go:34]: Do we need the forXDP parameter in runBpfUnitTest?  If not, I think better to revert in order to save a few changes.

File: bpf/ut/main_test.go [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2020 Tigera, Inc. All rights reserved.
+// Copyright (c) 2020-2021 Tigera, Inc. All rights reserved.
```
Comment (neiljerram) [inline bpf/ut/main_test.go:1]: Should revert this copyright change, when the file isn't changing in any other way.

**Review by neiljerram — CHANGES_REQUESTED** [review]
Looking good, thanks.  I think there are just a couple more kinds of reversion needed.

- I don't think we need `forXDP` parameter in `runBpfUnitTest` calls, so please can we revert those changes?

- After other reversions, when a file is not changing in any way apart from the copyright line, we should remove that file completely from the PR (aka leave its copyright line as it was before).  You may need the `git commit` `--no-verify` option for this, to skip our hooks when removing those files from the commit.

Then one nit: maybe move the XDP jump map definition into its own package?  I don't feel strongly about this though.

Finally: is it intended that you haven't added new XDP tests yet?  I guess they might come in the next PR, but it's also possible that you missed `git add` for that file!

File: bpf/ut/bpf_prog_test.go (hunk 3 above)
Comment (mazdakn) [reply bpf/ut/bpf_prog_test.go:314]: The only thing is that the xdpjump is only needed for ut. We don't need anywhere else. For that, I felt maybe it is better to define it somewhere else. Does it make sense?

Comment (neiljerram) [reply bpf/ut/bpf_prog_test.go:314]: OK, no worries, that's absolutely fine.

Comment (mazdakn) [reply bpf/ut/bpf_prog_test.go]: Well, just realised that the extra map is only needed for "runBpfUnitTest", so it is also removed now.

**Comment by mazdakn** [comment]
> Finally: is it intended that you haven't added new XDP tests yet? I guess they might come in the next PR, but it's also possible that you missed `git add` for that file!

Yes, I'll use another PR for the actual tests. The thing is that initially I worked on the "runBpfUnitTest" which later on looked useless for XDP. Then I focused on "runBpfTest", but that needs metadata to work. After this PR, I'll start to make "runBpfTest" tests like failsafe ones to work.

**Review by neiljerram — APPROVED** [review]
LGTM - thanks @mazdakn !

**Pull request merged**